Better stats and search in adherent module

daraelmin opened the pull request:
#Fix #13757 count too much entities or location and wrong history
Properties stats should count each adherent only once, not each subscription.
The status of the adherent isn't so important as the validation date. So it should be counted when, and only when, a subscription is up to date (paid). Historical stats should count each adherent who has subscribed in the past years even if the status of the adherent is retired.

#Fix BUG - Subscrition type may not be confirmed
Currently, the subscription view displays by default the adherent's current subscription type if no type has been registered in the past. If an update was made from an earlier version of Dolibarr (up to v10), the subscription type was not recorded in the tablellx_subscritipion. There is therefore nothing to display. This procedure is risky because it misleads the user into believing that the information is confirmed. In my opinion, it is better not to display anything as a type of subscription when the information does not exist. In certain circumstances, for legal and insurance matters, the history is particularly important. You should know if the adherents were covered several years back. If the information is missing on the platform, the user will know that he will have to retrieve this information otherwise, but if he sees information, he risks saying that it is true. It was far too risky.

#Fix Filter up/out-todate disappear when searching in adherent list
I just added an hidden field to keep the membership type (uptodate or outofdate) even in search mode. Particulary usefull when using tags/categories

daraelmin committed:
Adherents stats #13757 count too much entities

Properties stats should count each adherent only once, not each subscription.
The status of the adherent isn't so important as the validation date. So it should be counted when, and only when, a subscription is up to date (paid).

daraelmin committed:
Adherents stats #13757 count too much location

Location statistics should count each adherent only once, not each subscription.
The status of the adherent isn't so important as the validation date. So it should be counted when, and only when, a subscription is up to date (paid).

daraelmin committed:
Adherents stats #13757 Wrong historical count 

Historical stats should count each adherent who has subscribed in the past years even if the status of the adherent is retired.

daraelmin committed:
FIX BUG - Subscrition type is not confirmed

Currently, the subscription view displays by default the adherent's current subscription type if no type has been registered in the past.

If an update was made from an earlier version of Dolibarr (up to v10), the subscription type was not recorded in the tablellx_subscritipion. There is therefore nothing to display.

This procedure is risky because it misleads the user into believing that the information is confirmed. In my opinion, it is better not to display anything as a type of subscription when the information does not exist.

In certain circumstances, for legal and insurance matters, the history is particularly important. You should know if the adherents were covered several years back.

If the information is missing on the platform, the user will know that he will have to retrieve this information otherwise, but if he sees information, he risks saying that it is true.

It's far too risky.

daraelmin committed:
FIX|Fix Filter uptodate or outodate disappear when using search in adherent list

I just added an hidden field to keep the membership type (uptodate or outofdate) even in search mode 
particulary usefull when using tags/categories

daraelmin committed:
FIX/fix BUG - Subscrition type is not confirmed 



Currently, the subscription view displays by default the adherent's current subscription type if no type has been registered in the past.

If an update was made from an earlier version of Dolibarr (up to v10), the subscription type was not recorded in the tablellx_subscritipion. There is therefore nothing to display.

This procedure is risky because it misleads the user into believing that the information is confirmed. In my opinion, it is better not to display anything as a type of subscription when the information does not exist.

In certain circumstances, for legal and insurance matters, the history is particularly important. You should know if the adherents were covered several years back.

If the information is missing on the platform, the user will know that he will have to retrieve this information otherwise, but if he sees information, he risks saying that it is true.

It's far too risky.

daraelmin committed:
FIX BUG - Subscrition type may not be confirmed 

Currently, the subscription view displays by default the adherent's current subscription type if no type has been registered in the past.

If an update was made from an earlier version of Dolibarr (up to v10), the subscription type was not recorded in the tablellx_subscritipion. There is therefore nothing to display.

This procedure is risky because it misleads the user into believing that the information is confirmed. In my opinion, it is better not to display anything as a type of subscription when the information does not exist.

In certain circumstances, for legal and insurance matters, the history is particularly important. You should know if the adherents were covered several years back.

If the information is missing on the platform, the user will know that he will have to retrieve this information otherwise, but if he sees information, he risks saying that it is true.

It's far too risky.

eldy commented on htdocs/adherents/stats/geo.php:
```diff
@@ -74,12 +74,12 @@
         $sql .= " LEFT JOIN ".MAIN_DB_PREFIX."subscription as s ON s.fk_adherent = d.rowid";
         $sql .= " WHERE d.entity IN (".getEntity('adherent').")";
-        $sql .= " AND d.statut = 1";
+        $sql .= " AND CURDATE() BETWEEN s.dateadh AND s.datef";
```
what if s.datef is not defined ?
for example when no subscription was yet recorded. This seems to exclude such record.
Also CURDATE is not allowed since it uses the timezone of database that differs on timezone of application. Try to use php date dol_now().

daraelmin replied:
Un my opinion, if no subscription is recorded, then the man is a prospect, not a member not yet an adherent

eldy commented on htdocs/adherents/stats/geo.php:
```diff
@@ -90,14 +90,14 @@
         $sql .= " LEFT JOIN ".MAIN_DB_PREFIX."c_country as co on d.country = co.rowid";
         $sql .= " LEFT JOIN ".MAIN_DB_PREFIX."subscription as s ON s.fk_adherent = d.rowid";
         $sql .= " WHERE d.entity IN (".getEntity('adherent').")";
-        $sql .= " AND d.statut = 1";
```
By definition a "draft" object in dolibarr is an object that does not exist (it is a draft that may be deleted or validated to exists officially), so we should exclude them from statistitcs.

daraelmin replied:
OK, but for historical stats, and comparison, you should include the 
retired adherent, maybe de van check if there's a validation date 
(d.datevalid != "0000-00-00 00:00:00") ?
Otherwise it is a wrong info.

eldy replied:
Yes filtering on -1 is better. And yes, including the terminated member for stats is usefull. But it also break a feature used at end of year by association to know how many valid/active member there is (required for the annualy moral report).
So the only solution i see is to introduce a filter with 3 choices:
Empy choice
Open only
Terminated only

Filter empty -> All members (except draft). What you expect
Filter on open only status = 1 (like today). So we don't loose a feature.
Filter on terminated status = 0. New feature

It is more work, but the only solution i see to provide best of the 2 needs (current and the one you are looking for).

daraelmin commented:
Un m'y opinion, if no subscription ils recorded, then thé man is a prospect, not a membre.
Cordialement
Waël ALMOMAN

-- 
Envoyé de mon téléphone LineageOS, en vous priant donc d'excuser ma brièveté.

Le 27 août 2020 20:55:19 GMT+02:00, Laurent Destailleur <notifications@github.com> a écrit :
>@eldy commented on this pull request.
>
>
>
>>          $sql .= " FROM ".MAIN_DB_PREFIX."adherent as d";
>         $sql .= " LEFT JOIN ".MAIN_DB_PREFIX."c_country as c on d.country = c.rowid";
>         $sql .= " LEFT JOIN ".MAIN_DB_PREFIX."subscription as s ON s.fk_adherent = d.rowid";
>         $sql .= " WHERE d.entity IN (".getEntity('adherent').")";
>-        $sql .= " AND d.statut = 1";
>+        $sql .= " AND CURDATE() BETWEEN s.dateadh AND s.datef";
>
>what if s.datef is not defined ?
>for example when no subscription was yet recorded. This seems to exclude such record.
>Also CURDATE is not allowed since it uses the timezone of database that differs on timezone of application. Try to use php date dol_now().
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/Dolibarr/dolibarr/pull/14538#pullrequestreview-476995542

daraelmin committed:
Adherents stats #13757 Still wrong history 

Second proposal to exclude drafts, but count retired adherent in historical stats

daraelmin committed:
Adherents stats #13757 still too much location 

Second proposal with :
- dol_now() instead of CURDATE 
- counting valid adherent if there's no end subscription date.

daraelmin committed:
Adherents stats #13757 still too much entities

Second proposal with :
- dol_now() instead of CURDATE 
- counting valid adherent if there's no end subscription date.

daraelmin commented:
Hope this one is better to you.
Let me know.

daraelmin committed:
Delete end line white spaces

Just remove some white spaces to pass the code check

eldy commented:
Yes filtering on -1 is better. And yes, including the terminated member for stats is usefull. But it also break a feature used at end of year by association to know how many valid/active member there is (required for the annualy moral report).
So the only solution i see is to introduce a filter with 3 choices:
Empy choice
Open only
Terminated only

Filter empty -> All members (except draft). What you expect
Filter on open only status = 1 (like today). So we don't loose a feature.
Filter on terminated status = 0. New feature

It is more work, but the only solution i see to provide best of the 2 needs (current and the one you are looking for). Are you ok to enhance PR with this filter ?

daraelmin commented:
Ok 
Give two weeks. At the time I have work, but I'll work on it.

i'm not sure that we should use filter on subscription stats pages. Won't it be better with 3 columns : Actually Valid ; Actually terminated ; Total ?

daraelmin commented:
But what about the geo.php and properties php ? On these pages it is not specified whats are / should be the filter. In my mind it was nesserly the actual subscriber (and not necessarly the adherent) but... I'm not sur of any thing know. 

In fact I don't really need these info, I don't really know what should be the useful info.

eldy commented:
geo.php and properties php has the same troubles. We don't know what it returns (current only or from al time ?) So we should have a filter too. Or, like you suggest, we can also add 3 columns: Actually Valid ; Actually terminated ; Total
Adding a filter seems more simple because, it is just adding a combo and changing the where using the value.

Adding 3 columns is better. We have all what we may need in same page and it's easier to compare past/now. But it needs more work.

Do what is the best for you. Both choice are valid to be merged.

daraelmin commented:
In fact the probleme its not so similar cause geo.php count a property that refere to the adherent table, not the Subscription table. That means... We dont know where was living the adherent when the Subscription was bought.

I really think that the info showed actually is just wrong. If one adherent has suscribed during 10 years in Paris, but is now located in Marseille, it count 11 times Marseille.

daraelmin committed:
Add column to the stats

daraelmin committed:
Update adherentstats.class.php

daraelmin committed:
deleting white spaces at end line

daraelmin commented:
Ok I have add the two columns for the subscription stats. If you think it is ok, i'll try to add those on the widget. 

I still believe that it is a bug not a feature to list the location for the member of all time.
Tell me what you think we should do with the properties.php. Is it really relevant to list the nature of the subscribers of all time and to count them the number of times they have susribed ? I f'on't think so.
